working/static metaprogramming/feature_specification.md

```diff
@@ -0,0 +1,189 @@
+annotated with the macro.
+
+For example, lets look at the [ClassDeclarationMacro][ClassDeclarationMacro]
+API, which has the following method that you must override:
+
+```dart
+class ClassDeclarationMacro implements DeclarationMacro {
+  void visitClassDeclaration(
+      ClassDeclaration declaration, ClassDeclarationBuilder builder);
+}
+```
+
+The [ClassDeclaration][ClassDeclaration] instance you get here provides all the
+introspective information to you that is available for classes in the
+`declaration` phase.
+
+The `builder` instances also provide an api that allows you to retrieve an
+introspection object for any `Type` object available to the macro itself, so
+for instance the [DeclarationBuilder][DeclarationBuilder] class (which
+[ClassDeclarationBuilder][ClassDeclarationBuilder] extends), exposes the
+[typeDeclarationOf][typeDeclarationOf] api.
+
+### Code Building API
+
+At the root of the API for generating code, are the [Code][Code] and
+`*Builder` classes. The `*Builder` instance is always passed as the second
+argument to the methods you override in your macro.
+
+In your macro, you will generally use the introspection api you are given in
+order to construct the desired [Code][Code], and then pass that to an api from
+the builder instance in order to add it to the program.
+
+There is a different type of `*Builder` for each specific type of macro, for
+instance if you look at the [DeclarationBuilder][DeclarationBuilder] class, you
+will see this interface method `void addToLibrary(Declaration declaration)`.
+The [Declaration][Declaration] class here is a subtype of [Code][Code].
+
+Most subtypes of [Code][Code] require fully syntactically valid code in order to
+be constructed, but where you need to build up something in smaller pieces you
+can use the [Fragment][Fragment] subtype. Any arbitrary String can be passed to
+this class, allowing you to build up your code fragments however you like.
+
+## Scoping
+
+**TODO**: Fill in this section with more detail and a real proposal.
+
+Macros will likely want to introduce references to identifiers that are not in
+the scope of the library in which they are running, but are in the scope of the
+macro itself.
+
+We will want some way of providing an affordance to emit a reference to
+something from the macro scope, but in the code generated for the original
+library.
+
+We do have the start of something like this already available in the `builder`
+api - these have apis to get a reference to a `Type` object. We will want to add
+the ability to do the same for any arbitrary identifier, and then the ability
+to emit references to these inside of [Code][Code] objects.
+
+## Limitations
+
+- Macros cannot be applied from within the same library cycle as they are
+  defined.
+- Macros cannot write arbitrary files to disk, and read them in later. They
+  can only generate code into the library where they are applied.
```

Comment:
We should probably have a TODO here to figure out which "dart:" libraries and APIs can be used.